[AST][Type] Introduce ObjectType; Infer the type of call_packed by type_args; Refactor InferType/InferShape.

YuchenJin opened the pull request:
As discussed in https://github.com/tlc-pack/relax/issues/122, we need to ensure the type invariant that "All Expr should have not null checked_type_ after type inference" always hold.

This PR introduces `ObjectType`, which serves as a general type, and its runtime representations can be `tvm::runtime::Object` such as `String`, `Array`, `Integer`, etc, and can be used when we support TensorArray.

For the return type of a `call_packed` node, in most cases we know the return type of a packed function. For example, if a packed function performs a tensor operation, the return value should be of DynTensorType. This PR annotates the return type by the CallNode's `type_args`,  and adds a general handling for `type_args` in the parser and printer.

In TVMScript, user can annotate the return type of a call_packed as:
```python
# return a tensor
x = relax.call_packed("test.vm.mul", x, y, type_args=(Tensor(rank=2, dtype="float32")))`

# return a shape
y = relax.call_packed("contrib.my_shape_of", x, type_args=(Shape))`

# return an object
z = relax.call_packed("contrib.tensor_array_concat", x, y, type_args=(Object))`
```
If `type_args` is not annotated, by default the return type of `call_packed` would be `ObjectType`.

This PR also refactors the `InferType` and `InferShape` functions in the Normalizer, and added the FInferType/FInferShape attributes to relax ops to make sure we can infer the types for all ops.

tqchen commented on python/tvm/relax/ty.py:
```diff
@@ -24,10 +24,24 @@
 
+@tvm._ffi.register_object("relax.ObjectType")
+class ObjectType(Type):
+    """A general type in Relax.
```
Object type correspond to tvm::runtime::Object, is base of all possible object values in TVM.

tqchen commented on python/tvm/script/relax/parser.py:
```diff
@@ -1000,12 +1002,57 @@ def _get_values(expr: ast.ArrayLiteral, vals: List[Any]) -> List[Any]:
+        """
+        for key, val in expr.keyword_params.items():
+            if key.value == "type_args":
+                if isinstance(val, ast.Call):
```
readability: collect first then act

YuchenJin replied:
Refactored.

tqchen commented on python/tvm/script/relax/parser.py:
```diff
@@ -1000,12 +1002,57 @@ def _get_values(expr: ast.ArrayLiteral, vals: List[Any]) -> List[Any]:
+                    if val.func_name.id.name != "Tensor":
+                        self.report_error(
+                            f"type_args elements must be Tensor, Object, or Shape", val.span
+                        )
```
parse_type

YuchenJin replied:
Added parse_type_from_value.

tqchen commented on python/tvm/script/relax/parser.py:
```diff
@@ -1000,12 +1002,57 @@ def _get_values(expr: ast.ArrayLiteral, vals: List[Any]) -> List[Any]:
+                            f"type_args elements must be Tensor, Object, or Shape", val.span
+                        )
+                elif isinstance(val, ast.Tuple):
+                    fields = [self.parse_call_type_args(field) for field in val.values]
```
self, parse_type?

YuchenJin replied:
Regression test added.

tqchen commented on python/tvm/script/relax/parser.py:
```diff
@@ -1000,12 +1002,57 @@ def _get_values(expr: ast.ArrayLiteral, vals: List[Any]) -> List[Any]:
+                elif isinstance(val, ast.Var):
+                    if val.id.name == "Object":
+                        return relax.ObjectType(self.to_tvm_span(val.span))
+                    elif val.id.name == "Shape":
```
regression

tqchen commented on src/relax/backend/vm/vm_shape_lower.cc:
```diff
@@ -119,6 +118,11 @@ class VMShapeLowerMutator : public ExprMutator {
+    Type ret_type = this->VisitType(node->ret_type);
+    // weaken the ret_type given this pass is the last stage of build.
+    if (const DynTensorTypeNode* temp = ret_type.as<DynTensorTypeNode>()) {
+      ret_type = DynTensorType(-1, temp->dtype);
```
DynTensorType::kUnknownRank,

DynTensorType::UnknownNDim(dtype)

tqchen replied:
document the behavior, change all tensor types to unknown rank

YuchenJin replied:
Updated.

tqchen commented on src/relax/ir/block_builder.cc:
```diff
@@ -366,8 +365,12 @@ class BlockBuilderNode::ExprNormalizer : public ExprFunctor<Expr(const Expr&)> {
+      return RuntimeDepShape(Span());
+    } else if (call->op.as<OpNode>()) {
+      // primitive op: look up FInferShape attribute
+      auto op_map = Op::GetAttrMap<FInferShape>("FInferShape");
```
static auto&

YuchenJin replied:
Done.

tqchen commented on src/relax/ir/block_builder.cc:
```diff
@@ -379,34 +382,56 @@ class BlockBuilderNode::ExprNormalizer : public ExprFunctor<Expr(const Expr&)> {
         }
       }
+    } else {
+      LOG(FATAL) << "ValueError: Failed to do shape inference.";
```
for call->op->GetType();

tqchen commented on src/relax/ir/block_builder.cc:
```diff
@@ -379,34 +382,56 @@ class BlockBuilderNode::ExprNormalizer : public ExprFunctor<Expr(const Expr&)> {
-    if (call->op.as<OpNode>()) {
+    // call_packed: return type_args of the call node
+    if (call->op.as<ExternFuncNode>()) {
+      if (call->type_args.defined()) {
```
TODO: add type annotation for ExternFuncNode

tqchen commented on src/relax/ir/block_builder.cc:
```diff
@@ -379,34 +382,56 @@ class BlockBuilderNode::ExprNormalizer : public ExprFunctor<Expr(const Expr&)> {
+      }
+    } else if (call->op.as<OpNode>()) {
+      // primitive op: look up FInferType attribute
+      auto op_map = Op::GetAttrMap<FInferType>("FInferType");
```
cache finfer_type

YuchenJin replied:
made it a member.

tqchen commented on src/relax/ir/block_builder.cc:
```diff
@@ -379,34 +382,56 @@ class BlockBuilderNode::ExprNormalizer : public ExprFunctor<Expr(const Expr&)> {
       if (it_func != ctx_mod->functions.end()) {
         if (const auto* func = (*it_func).second.as<FunctionNode>()) {
           return func->checked_type_;
         }
```
Add TODO(yuchen) check after normalization

tqchen replied:
TODO(yuchen) call to local var

tqchen commented on src/relax/op/op.cc:
```diff
@@ -51,6 +51,12 @@ bool EqualCheck(const PrimExpr& lhs, const PrimExpr& rhs) {
   return false;
 }
 
+Type InferTypeVoid(const Call& call, DiagnosticContext diag_ctx) { return VoidType(); }
```
ReturnVoidType

YuchenJin commented:
@tqchen, thanks for the review and great suggestions! I addressed your comments, could you take a second pass? Also I made the type_args to be required in call_packed.

tqchen commented:
Thank you @YuchenJin !
